src/middlewared/middlewared/plugins/gluster_/peer.py

```diff
@@ -0,0 +1,122 @@
+import asyncio
+from gfs.cli import peer
+from gfs.cli.utils import GlusterCmdException
+
+from middlewared.async_validators import resolve_hostname
+from middlewared.schema import Dict, Str
+from middlewared.service import (accepts, private, job,
+                                 CallError, CRUDService,
+                                 ValidationErrors)
+
+
+class GlusterPeerService(CRUDService):
+
+    class Config:
+        namespace = 'gluster.peer'
+
+    def __peer_wrapper(self, method, host=None):
+
+        result = ''
+        try:
+            result = method(host) if host else method()
+        except GlusterCmdException as e:
+            rc, out, err = e.args[0]
+            err = err if err else out
+            raise CallError(f'{err.decode().strip()}')
+        except Exception:
+            raise
+
+        if isinstance(result, bytes):
+            return result.decode().strip()
+
+        return result
+
+    @private
+    def remove_peer_from_cluster(self, hostname):
+
+        return self.__peer_wrapper(peer.detach, host=hostname)
+
+    @private
+    def add_peer_to_cluster(self, hostname):
+
+        return self.__peer_wrapper(peer.attach, host=hostname)
+
+    @private
+    async def resolve_host_or_ip(self, hostname, verrors):
+
+        args = (self.middleware, verrors, 'resolve_host_or_ip', hostname)
+        return await resolve_hostname(*args)
+
+    @private
+    def common_validation(self, hostname):
+
+        verrors = ValidationErrors()
+
+        loop = self.middleware.loop
+        try:
+            asyncio.run_coroutine_threadsafe(
```

Comment:
You can do `self.middleware.call_sync('gluster.peer.resolve_host_or_ip')`, that's what it is there for, to be able to switch between threads/asyncio easily

Reply:
Thank you!